### PR TITLE
Add zones to an existing row

### PR DESCRIPTION
**Pull request by chelliah**

### CODE REVIEW

**Comment by chelliah** [comment]
UI for reordering could use some work

**Comment by chelliah** [comment]
updated 
now you can:
move zones to a new row
move zones to a new or existing column in any existing row

File: src/components/EditorWrapper.js [hunk 1]
```diff
@@ -106,15 +92,6 @@ export default class EditorWrapper extends React.Component {
     return (
       <div name="EditorWrapper" className="zone-content" ref={(el) => this.wrapper = el}>
         <div className={ containerClass }>
-          { isHover &&
-            <div style={hoverButtonStyles}>
-              <EditButton
```
Comment (rayd) [inline src/components/EditorWrapper.js]: If you remove this, how do we ever start editing?

File: src/components/RowContainer.js [hunk 2]
```diff
@@ -1,47 +1,215 @@
...
+            } }>
+            <MoveVertButton
+              smallButton
+              color='#FFAA39'/>
+            </section>)
+      }
+      <Row
+        {...this.props}
+      />
+      <DragZone
```
Comment (rayd) [inline src/components/RowContainer.js]: Should we really have these `DragZone`s on either end of rows and in-between all zones in the row, so users can insert their dragged zone anywhere within a row, without having to swap with an existing zone?

File: src/components/RowContainer.js [hunk 3]
```diff
@@ -1,47 +1,215 @@
...
+        moveZoneToNewColumn={ this.moveZoneToNewColumn }
+        rowId={row.get('id')}
+        isHoveringOverContainer={isHoveringOverContainer} />
+      <AddButtonHorizRule
```
Comment (rayd) [inline src/components/RowContainer.js]: Same thing goes for this -- I'd love to be able to just create a new zone in between two existing zones. Should this button be present on either side of every single zone?

File: src/reducers/zones.js [hunk 4]
```diff
@@ -1,5 +1,5 @@
 import Actions from '../helpers/actionConstants';
-import { Map } from 'immutable';
+import { Map, List } from 'immutable';
```
Comment (rayd) [inline src/reducers/zones.js]: doesn't look like we need this

File: src/components/EditorWrapper.js (hunk 1 above)
Comment (chelliah) [reply src/components/EditorWrapper.js]: click anywhere on the component, the 'edit button' has been a bit redundant for a while

File: src/reducers/rows.js [hunk 5]
```diff
@@ -41,6 +41,29 @@ export default function rows(state = List(), action) {
...
+      newState = newState
+        .map((row) => {
+          if (row.get('id') === action.rowId) {
+            const zones = row.get('zones');
+            return row.set('zones', zones.push(action.zone));
+          }
+          return row;
+        });
+      break;
+    case Actions.EDITOR_INSERT_ZONE:
```
[the remark this answers is not in the document]
Comment (chelliah) [reply src/reducers/rows.js]: true

File: src/components/RowContainer.js (hunk 3 above)
Comment (chelliah) [reply src/components/RowContainer.js]: yeah that'd be nice but its mostly just like a UI/space allocation issue, lots of potentially overlapping elements

File: src/components/RowContainer.js (hunk 2 above)
Comment (chelliah) [reply src/components/RowContainer.js]: so thats kind of how it works currently, in order to drag to the beginning or in between any zones you hover over the existing components, theres a line to tell you where the dragging component will go

Comment (chelliah) [reply src/components/RowContainer.js]: the drag zone is really just an empty container to handle a drop in a new space

File: src/components/Canvas.js [hunk 6]
```diff
@@ -283,6 +292,60 @@ export class Canvas extends React.Component {
     this.addRow('Image', rowsToAdd);
   }
 
+  addZone(type, rowId, defaultAction, existingProps = {}) {
```
[the remark this answers is not in the document]
Comment (chelliah) [reply src/components/Canvas.js]: yeah theres a ton of repeated logic here..

File: src/components/AddButtonHorizRule.js [hunk 7]
```diff
@@ -86,39 +138,98 @@ export default class AddButtonHorizRule extends React.Component {
...
     if (isHoveringOverContainer || showEditorSelector) {
       //alter styles whenever a users mouse is over any part of the wysiwyg editor
       hrStyle = {
         ...hrStyle,
         opacity: 0.5
       };
 
+      addButtonStyle = {
+        ...addButtonStyle,
+        pointerEvents: 'all'
```
[the remark this answers is not in the document]
Comment (chelliah) [reply src/components/AddButtonHorizRule.js]: iirc its beause the green line/box that shows up on hover overlapped with the skip and next buttons

[the remark this answers is not in the document]
Comment (chelliah) [reply src/components/AddButtonHorizRule.js]: youre right though, its a pretty bad practice

File: src/components/RowContainer.js (hunk 2 above)
Comment (chelliah) [reply src/components/RowContainer.js]: Drag zones might be a good idea going forward though, the drag and drop feels kind of weird at the moment

File: src/components/AddButtonHorizRule.js [hunk 8]
```diff
@@ -86,39 +138,98 @@ export default class AddButtonHorizRule extends React.Component {
...
+          };
+
+          containerStyle = {
+            ...containerStyle,
+            zIndex: 110,
+            width: ADD_ZONE_EXPANDED,
+          };
+        }
+    } else {
+      if (isHoveringOverAddButton || showEditorSelector  || (canDrop && isOver)) {
```
[the remark this answers is not in the document]
Comment (chelliah) [reply src/components/AddButtonHorizRule.js]: i think its actually nested in the (isHorizontalOrientation) part of the statement, the whole styling here is a bit of a mess bc doing everything with inline styles is so tricky

File: src/components/AddButtonHorizRule.js [hunk 9]
```diff
@@ -136,19 +247,20 @@ export default class AddButtonHorizRule extends React.Component {
...
+            onClick={ this.handleAddNew }
             onMouseOver={() => !isHoveringOverAddButton && this.setState({ isHoveringOverAddButton: true })}
             onMouseOut={() => isHoveringOverAddButton && this.setState({ isHoveringOverAddButton: false })}
             ref={(el) => this.addButton = el}
           >
             <AddButtonContainer
               onSelectEditorType={ onSelectEditorType }
-              onClick={ this.handleAddNew }
               showEditorSelector={ showEditorSelector }
+              smallButton={ !isHorizontalOrientation }
```
[the remark this answers is not in the document]
Comment (chelliah) [reply src/components/AddButtonHorizRule.js]: it is, just passed through indirectly via ...this.props

[the remark this answers is not in the document]
Comment (rayd) [reply src/components/AddButtonHorizRule.js]: oops :grimacing:

**Comment by GuruM** [comment]
if it's not too much a bother, could you post an updated gif?

**Pull request merged**